### PR TITLE
Add support for highlighting many matches at once

### LINKED ISSUE #7: bug:
### Did you check docs and existing issues?

- [X] I have read all the paint.nvim docs
- [X] I have searched the existing issues of paint.nvim
- [X] I have searched the exsiting issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.8.2

### Operating system/version

ArchLinux 6.1.4-zen2-1-zen

### Describe the bug

Paint doesn't match the same pattern multiple times per line

For example the patter `hi` will only match the first `"hi"` in `"hi hey hi"`

### Steps To Reproduce

...

### Expected Behavior

...

### Repro

```Lua
...
```

### PR DESCRIPTION
**Pull request by B4mbus**
Added support for highlighting many matches at once. Some people may not need to so I can make it an option, for now I leave it how it is.

Fixes #7 
Fixes #3

### CODE REVIEW

File: lua/paint/highlight.lua [hunk 1]
```diff
@@ -37,6 +39,9 @@ function M.highlight(buf, first, last)
           from - 1,
           { end_col = to, hl_group = hl.hl, priority = 110 }
         )
+
+        last_to = to
```
Comment (folke) [inline lua/paint/highlight.lua:43]: best to make this `to+1`, just to be sure we don't enter an endless loop. Could probably only happen when highlighting a single character, but still :)

**Review by folke — CHANGES_REQUESTED** [review]

File: lua/paint/highlight.lua (hunk 1 above)
Comment (B4mbus) [reply lua/paint/highlight.lua]: Right, did that, but I still have an infinite loop somewhere (that's why I initially marked it as draft), I'm trying to debug it.

File: lua/paint/highlight.lua [hunk 2]
```diff
@@ -25,8 +25,10 @@ function M.highlight(buf, first, last)
     local lnum = first + l - 1
 
     for _, hl in ipairs(highlights) do
-      local from, to, match = line:find(hl.pattern)
-      if from then
+      local last_to = 1
+      local from, to, match = line:find(hl.pattern, last_to)
```
Comment (folke) [inline lua/paint/highlight.lua:29]: this needs to move in the while loop

File: lua/paint/highlight.lua [hunk 3]
```diff
@@ -25,8 +25,10 @@ function M.highlight(buf, first, last)
...
+
+      while from do
```
Comment (folke) [inline lua/paint/highlight.lua:31]: make this a while true

File: lua/paint/highlight.lua [hunk 4]
```diff
@@ -25,8 +25,10 @@ function M.highlight(buf, first, last)
...
         if match then
```
Comment (folke) [inline lua/paint/highlight.lua:32]: add that previous line I mentioned right above this and when not match, break

**Comment by B4mbus** [comment]
Okay. I fixed the infinite loop and it seems to work correctly with every single highlight I have. To be fair I don't really understand what you want me to do in the code comments, can you elaborate?

File: lua/paint/highlight.lua [hunk 5]
```diff
@@ -25,18 +25,24 @@ function M.highlight(buf, first, last)
     local lnum = first + l - 1
 
     for _, hl in ipairs(highlights) do
-      local from, to, match = line:find(hl.pattern)
-      if from then
-        if match then
+      local last_to = 1
+      local from, to, match = line:find(hl.pattern, last_to)
```
Comment (folke) [inline lua/paint/highlight.lua:29]: Just pass 1 here and remove the `last_to` variable

**Review by folke — CHANGES_REQUESTED** [review]
Two tiny changes and all good to go!
I think you need to change the draft status though, so that I can merge this.
Thanks!

**Comment by folke** [comment]
Thank you!

**Pull request merged**